code/game/objects/items/stacks/stack.dm

```diff
@@ -463,6 +463,10 @@
 	target_stack.copy_evidences(src)
 	use(transfer, transfer = TRUE, check = FALSE)
 	target_stack.add(transfer)
+	if(target_stack.mats_per_unit != src.mats_per_unit)
+		var/M
+		for(M = 1, M < target_stack.mats_per_unit, M++)
```

Comment:
You don't need to put the var/M outside of the loop, and you also don't really need to do this c style loop.

`for(var/mat_type in target_stack.mats_per_unit)`
will do just fine

Oh also you'll wanna avoid things like single letter var names, since the goal is to make the code as easy as possible for other people to read/understand

Having var names that do something to describe their purpose helps a lot with this

Reply:
for(var/mat_type in target_stack.mats_per_unit)
i tried this and instead of 1,2 it give Iron, Glass

Reply:
Ah lemme be more precise 

`target_stack.mats_per_unit[target_stack.mats_per_unit[M]] = (target_stack.mats_per_unit[target_stack.mats_per_unit[M]] * target_stack.amount + src.mats_per_unit[src.mats_per_unit[M]] * transfer) / (target_stack.amount + transfer)`

In this big line you're doing a lot of `stack.mats[stack.mats[M]]`.
`mats_per_unit` or mats is an associated list. That means it looks like this
"Key" -> Value

What you're currently doing is going over the list by index, using the index M to get the list's key, and then using that key to get the value.

So M is the index, `stack.mats[M]` is the key, and `stack.mats[stack.mats[M]]` is the value.

Byond lets you do for loops like this `for(var/key in assoc_list)`
So instead of getting the index, then the key, then the value, we can just get the key and use it to set/get the value.

So that big block of code would become
`target_stack.mats_per_unit[mat_type] = (target_stack.mats_per_unit[mat_type ] * target_stack.amount + src.mats_per_unit[mat_type ] * transfer) / (target_stack.amount + transfer)`

That make any sense?

By the by you don't need the src.mats_per_unit stuff either, since just mats_per_unit will be treated the same.

Reply:
oh hold i think i get what you trying to say, let me check